### PR TITLE
TxBuilder: TransactionBuilder.nodeStake takes rewardDelegators

### PR DESCRIPTION
**Pull request by msmania**
This patch implements a new feature Reward Delegators ([PIP-32](https://forum.pokt.network/t/pip32-unleashing-the-potential-of-non-custodial-node-running/4796); introduced by https://github.com/pokt-network/pocket-core/pull/1581).

### CODE REVIEW

File: packages/transaction-builder/src/models/msgs/msg-proto-node-stake.ts [hunk 1]
```diff
@@ -30,14 +31,16 @@ export class MsgProtoNodeStakeTx extends TxMsg {
     outputAddress: string,
     chains: string[],
     amount: string,
-    serviceURL: URL
+    serviceURL: URL,
```
Comment (nodiesBlade) [inline packages/transaction-builder/src/models/msgs/msg-proto-node-stake.ts:35]: add jsdoc param for delegators

**Comment by msmania** [comment]
Here's the sample code to add delegators.  Tx is [this](https://v2.poktscan.com/testnet/tx/A167CE9318FB41F32014FE238E776FB5FACC2D4C28B69CB8D6BDE2887C9D7CB2).  Will update README.md in a follow-up patch.

```js
import "dotenv/config";
import { JsonRpcProvider } from "@pokt-foundation/pocketjs-provider";
import { KeyManager } from "@pokt-foundation/pocketjs-signer";
import { TransactionBuilder } from "@pokt-foundation/pocketjs-transaction-builder";

const PoktEndpoint = process.env.POKT_ENDPOINT;
const TxSignerKey = process.env.SIGNER_PRIVATE_KEY;

async function main() {
  const provider = new JsonRpcProvider({
    rpcUrl: PoktEndpoint,
    dispatchers: [PoktEndpoint],
  });

  const height = await provider.getBlockNumber();
  console.log(height);

  const txSigner = await KeyManager.fromPrivateKey(TxSignerKey);
  const transactionBuilder = new TransactionBuilder({
    provider,
    signer: txSigner,
    chainID: "testnet",
  });
  const sendMsg = transactionBuilder.nodeStake({
    amount: "16000000000",
    chains: ["0001", "0002"],
    serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
    outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c",
    rewardDelegators: {
      "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c": 30,
      "54751ae3431c015a6e24d711c9d1ed4e5a276479": 40,
    }
  });
  const txResponse = await transactionBuilder.submit({
    memo: "via PocketJS",
    txMsg: sendMsg,
  });
  console.log(txResponse.txHash);
}

main()
  .then(() => process.exit(0))
  .catch(error => {
    console.error(error);
    process.exit(1);
  });
```

**Comment by nodiesBlade** [comment]
> Here's the sample code to add delegators. Tx is [this](https://v2.poktscan.com/testnet/tx/A167CE9318FB41F32014FE238E776FB5FACC2D4C28B69CB8D6BDE2887C9D7CB2). Will update README.md in a follow-up patch.
> 
> ```js
> import "dotenv/config";
> import { JsonRpcProvider } from "@pokt-foundation/pocketjs-provider";
> import { KeyManager } from "@pokt-foundation/pocketjs-signer";
> import { TransactionBuilder } from "@pokt-foundation/pocketjs-transaction-builder";
> 
> const PoktEndpoint = process.env.POKT_ENDPOINT;
> const TxSignerKey = process.env.SIGNER_PRIVATE_KEY;
> 
> async function main() {
>   const provider = new JsonRpcProvider({
>     rpcUrl: PoktEndpoint,
>     dispatchers: [PoktEndpoint],
>   });
> 
>   const height = await provider.getBlockNumber();
>   console.log(height);
> 
>   const txSigner = await KeyManager.fromPrivateKey(TxSignerKey);
>   const transactionBuilder = new TransactionBuilder({
>     provider,
>     signer: txSigner,
>     chainID: "testnet",
>   });
>   const sendMsg = transactionBuilder.nodeStake({
>     amount: "16000000000",
>     chains: ["0001", "0002"],
>     serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
>     outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c",
>     rewardDelegators: {
>       "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c": 30,
>       "54751ae3431c015a6e24d711c9d1ed4e5a276479": 40,
>     }
>   });
>   const txResponse = await transactionBuilder.submit({
>     memo: "via PocketJS",
>     txMsg: sendMsg,
>   });
>   console.log(txResponse.txHash);
> }
> 
> main()
>   .then(() => process.exit(0))
>   .catch(error => {
>     console.error(error);
>     process.exit(1);
>   });
> ```

Can you provide an example without delegators (basically before the addition as its an optional param).

----

As well, if a user doesn't specify delegators and there are existing delegators within a node stake, does that remove the delegators or keep it the same?

Example, are these two the same:

(is this even valid?)
```
  const sendMsg = transactionBuilder.nodeStake({
    amount: "16000000000",
    chains: ["0001", "0002"],
    serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
    outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c",
    rewardDelegators: {})
```

```
    const sendMsg = transactionBuilder.nodeStake({
    amount: "16000000000",
    chains: ["0001", "0002"],
    serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
    outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c")
```

**Comment by msmania** [comment]
> Can you provide an example without delegators (basically before the addition as its an optional param).
> 
> As well, if a user doesn't specify delegators and there are existing delegators within a node stake, does that remove the delegators or keep it the same?
> 
> Example, are these two the same:
> 
> (is this even valid?)
> 
> ```
>   const sendMsg = transactionBuilder.nodeStake({
>     amount: "16000000000",
>     chains: ["0001", "0002"],
>     serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
>     outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c",
>     rewardDelegators: {})
> ```
> 
> ```
>     const sendMsg = transactionBuilder.nodeStake({
>     amount: "16000000000",
>     chains: ["0001", "0002"],
>     serviceURL: new URL("https://poktt1698102386.c0d3r.org:443"),
>     outputAddress: "fc1c79e7efecf89f071ebb2ba7c6f5a98dcdfc3c")
> ```

Good question.

If we do `rewardDelegators: {}`, tx submission fails due to verification failure before transaction fee is deducted.

To stake without delegators or remove the current delegators, you just omit `rewardDelegators` or specify a falsy value like `undefined` or `null`.  I updated the comment on `nodeStake` about it.

In a follow-up patch, I'll add both scenarios, to add/remove delegators.

**Review by nodiesBlade — APPROVED** [review]
LGTM. As mentioned in other PR's and Discord, we will do a regression test of basic transactions of transfer, app stake/transfer, and staking using jsonrpcprovider and the webprovider to ensure all the commits did not affect functionality

**Pull request merged**